Resets session fail count on reload

aliking opened the pull request:
Hi, this fixes an issue that I'm seeing, but want to be sure I'm thinking of other people's cases.

I've set up an example that is similar to my use case here: https://github.com/aliking/webdriverio-browserstack

I'm using mocha bdd on browserstack, and I'm using a hook to reload before every example in my spec files, so that each has its own session. 

If I run a test that has multiple examples in it, any examples that run after a fail also get marked failed on Browserstack. 

That example has two spec files, each with two trivial examples: fail_pass.js has one example that fails, then one that passes. The other is called pass_fail.js (you get the idea). 

When I run it, my local reporting of the tests is correct; 2 passes, 2 fails. But on browserstack, there are 3 fails. 

This seems like the simplest way to fix my issue. Can you think of other cases where you wouldn't want to reset the failure count on reload?

aliking committed:
Resets session fail count on reload

itszero commented:
Thanks for reporting it. I need to wake up super early tomorrow so I won't be able to get to it right now. I'll definitely get to review this and test it out within this week. I currently do spot one error on the code: `.then(this.failures = 0)` would cause the statement run too early (i.e. not properly chained up in promise), but anyway, I'll check it out and fix that small bug.

Thanks again for helping out on this project! 👍

aliking committed:
Adds tests, makes them pass

aliking committed:
Merge remote-tracking branch 'upstream/master'

aliking committed:
Adds function order call test.

aliking commented:
Hi, thanks for looking. I realized how broken this PR was, so I've taken another stab at it. These commits resolve the conflicts that were in there and add some tests around what I'm trying to do and the functionality that I hope not to break. 

The changes in the actual service that I'd note are:
* I'm using `request.get` and `request.put` instead of just `request` so that I can stub out those functions in the test. 
* I reverted back the change at https://github.com/itszero/wdio-browserstack-service/pull/15/commits/212a6ef969d80f53a75128754caf65a267b46e62#diff-a502eea1ceec9a8bb68e6d6c866bd70eL26  - I'm not using that function, so there's not reason for it to be in there. 
* The call to `this._printSessionURL()` is no longer wrapped in a function. I think that _should_ work, and the tests I've added indicate that it seems to.

aliking commented:
Bump. I think the updates I've made to this are good - I believe the new unit tests are valid, and they work with my setup. Are you able to look at this? Hit me up with any questions.

itszero commented on lib/browserstack-service.js:
```diff
@@ -32,21 +32,29 @@ class BrowserstackService {
   }
 
   onReload(oldSessionId, newSessionId) {
+
```
please remove the unnecessary empty line :)

itszero commented on lib/browserstack-service.js:
```diff
@@ -32,21 +32,29 @@ class BrowserstackService {
     return this._update(oldSessionId, this._getBody())
-      .then(() => this._printSessionURL());
+      .then(this._printSessionURL())
+      .then(this._reset_failures());
```
Since `_reset_failures` is only used once, let's do: `.then(() => { this.failures = 0; })`

itszero commented on test/service.js:
```diff
@@ -0,0 +1,117 @@
+describe('BrowserstackService', function() {
+	const browserstack = new Browserstack;
+	before(function(){
+
```
remove

itszero commented on test/service.js:
```diff
@@ -0,0 +1,117 @@
+				'browserName': 'chrome'
+			}
+		}
+
```
can you please set your editor to 2 spaces and reindent this file? Thanks!

itszero commented on test/service.js:
```diff
@@ -0,0 +1,117 @@
+var sinon      = require('sinon'),
```
The style of the project regarding the imports are:
```
const sinon = require('sinon');
const request = require('request-promise');
...
```

itszero commented on test/service.js:
```diff
@@ -0,0 +1,117 @@
+var sinon      = require('sinon'),
+request = require('request-promise'),
+Promise = require('bluebird'),
```
please remove this. We use node.js native Promises.

itszero commented on test/service.js:
```diff
@@ -0,0 +1,117 @@
+Promise = require('bluebird'),
+Browserstack = require('../lib/browserstack-service.js');
+
+var getStub;
```
since we're using ES6 here, I would prefer these are `let`s :)

itszero commented on test/service.js:
```diff
@@ -0,0 +1,117 @@
+var putStub;
+
+describe('BrowserstackService', function() {
+	const browserstack = new Browserstack;
```
nit: `new Browserstack();`

itszero commented on test/service.js:
```diff
@@ -0,0 +1,117 @@
+	});
+
+	after(function(){
+		request.put.restore;
```
**IMPORTANT**: I think you'll need to call the restore function right? `request.put.restore();` There are a couple of these in this file.

itszero commented on test/service.js:
```diff
@@ -0,0 +1,117 @@
+
+	describe('#onReload()', function() {
+		before(function(){
+			sinon.spy(browserstack,'_update');
```
nit: style `sinon.spy(browserstack, '_update');` -- no spaces around parenthesis, space after comma. there're also a couple different styles in this file. let's unite them :)

itszero reviewed (changes requested):
Hey! Thanks for your enthusiasm! I've reviewed your code and left some comments. Let's get these bug/styling issues fixed and then we can merge your PR. 👍

aliking committed:
Address PR comments

aliking committed:
More extra lines

aliking commented:
Awesome, good feedback, thanks for this. All addressed.

itszero commented:
Merged. The release will have to wait until I added tests to both mocha/cucumber since we had some issues sharing the same code between both test frameworks.

Thanks for the contribution!
